pkg/helpers/imageregistry/client.go

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Red Hat, Inc.
+// Copyright Contributors to the Open Cluster Management project
+
+package imageregistry
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+const (
+	// ClusterImageRegistryLabel value is namespace.managedClusterImageRegistry
+	ClusterImageRegistryLabel = "open-cluster-management.io/image-registry"
+
+	// ClusterImageRegistriesAnnotation value is a json string of ImageRegistries
+	ClusterImageRegistriesAnnotation = "open-cluster-management.io/image-registries"
+)
+
+type Registries struct {
+	// Mirror is the mirrored registry of the Source. Will be ignored if Mirror is empty.
+	// +kubebuilder:validation:Required
+	// +required
+	Mirror string `json:"mirror"`
+
+	// Source is the source registry. All image registries will be replaced by Mirror if Source is empty.
+	// +optional
+	Source string `json:"source"`
+}
+
+// ImageRegistries is value of the image registries annotation
+type ImageRegistries struct {
+	PullSecret string       `json:"pullSecret"`
+	Registries []Registries `json:"registries"`
+}
+
+type Interface interface {
+	Cluster(cluster *clusterv1.ManagedCluster) Interface
+	PullSecret() (*corev1.Secret, error)
+	ImageOverride(imageName string) string
+}
+
+type Client struct {
+	kubeClient      kubernetes.Interface
+	imageRegistries ImageRegistries
+}
+
+func NewClient(kubeClient kubernetes.Interface) Interface {
+	return &Client{
+		kubeClient:      kubeClient,
+		imageRegistries: ImageRegistries{},
+	}
+}
+
+func (c *Client) Cluster(cluster *clusterv1.ManagedCluster) Interface {
+	annotations := cluster.GetAnnotations()
+	if len(annotations) == 0 {
+		c.imageRegistries = ImageRegistries{}
+		return c
+	}
+
+	_ = json.Unmarshal([]byte(annotations[ClusterImageRegistriesAnnotation]), &c.imageRegistries)
+	return c
+}
+
+func (c *Client) PullSecret() (*corev1.Secret, error) {
+	if c.imageRegistries.PullSecret == "" {
+		return nil, nil
+	}
+	segs := strings.Split(c.imageRegistries.PullSecret, ".")
+	if len(segs) != 2 {
+		return nil, fmt.Errorf("wrong pullSecret format %v in the annotation %s",
+			c.imageRegistries.PullSecret, ClusterImageRegistriesAnnotation)
+	}
+	namespace := segs[0]
+	pullSecret := segs[1]
+	return c.kubeClient.CoreV1().Secrets(namespace).Get(context.TODO(), pullSecret, metav1.GetOptions{})
+}
+
+func (c *Client) ImageOverride(imageName string) string {
+	if len(c.imageRegistries.Registries) == 0 {
+		return imageName
+	}
+	overrideImageName := imageName
+	for i := 0; i < len(c.imageRegistries.Registries); i++ {
```

Comment:
`for range` is simpler？

Reply:
just want to make it clean, the order can affect the results.

Reply:
let me add more doc here

Reply:
done